Toolkit/ArcGISToolkit/TemplatePickerViewController.swift

```diff
@@ -0,0 +1,303 @@
+        // if filtering, need to disable it
+        if isFiltering{
+            navigationItem.searchController?.isActive = false
+        }
+        
+        // add to list of unfiltered infos
+        unfilteredInfos = infos
+        
+        // re-assign to the current infos so we can update the tableview
+        currentInfos = unfilteredInfos
+        
+        // generate swatches for the layer infos
+        for index in infos.indices{
+            let info = infos[index]
+            if let feature = info.featureTable.createFeature(with: info.featureTemplate){
+                let sym = info.featureLayer.renderer?.symbol(for: feature)
+                sym?.createSwatch{ [weak self] image, error in
+                    
+                    guard let self = self else { return }
+                    guard error == nil else{ return }
+                    
+                    // update info with swatch
+                    infos[index].swatch = image
+                    
+                    // reload index where that info currently is
+                    let indexPath = self.indexPath(for: info)
+                    self.tableView.reloadRows(at: [indexPath], with: .automatic)
+                }
+            }
+        }
+    }
+    
+    // MARK: TableView delegate/datasource methods
+    
+    public func numberOfSectionsInTableView(_ tableView: UITableView) -> Int{
+        return tables.count
+    }
+    
+    public func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
+        guard !tables.isEmpty else { return nil }
+        return tables[section].tableName
+    }
+    
+    public func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        
+        // when the user taps on a feature type
+        
+        // first get the selected object
+        let selectedFeatureTemplateInfo = info(for: indexPath)
+        
+        // Note: we can't do this before the above, or else the object we pull
+        // out of the datasource will be incorrect
+        //
+        // If the search controller is still active, the delegate will not be
+        // able to dismiss this if they showed this modally
+        // (or wrapped it in a navigation controller and showed that modally)
+        // Only do this if not being presented from a nav controller
+        // as in that case, it causes problems when the delegate that pushed this VC
+        // tries to pop it off the stack.
+        if presentingViewController != nil{
+            navigationItem.searchController?.isActive = false
+        }
+        
+        delegate?.templatePickerViewController(self, didSelect: selectedFeatureTemplateInfo)
+    }
+    
+    override public func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        guard !tables.isEmpty else { return 0 }
+        let tableName = tables[section].tableName
+        let infos = self.currentDatasource[tableName]
```

Comment:
Another way of doing this would be to use the subscript variant that takes a default value:

```swift
return currentDatasource[tableName, default: []].count
```

Reply:
oh nice!